src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs

```diff
@@ -51,142 +34,177 @@ private ExpressionSyntax ParseInterpolatedStringToken()
 
-            var isAltInterpolatedVerbatim = originalText.Length > 2 && originalText[0] == '@'; // @$
-            var isVerbatim = isAltInterpolatedVerbatim || (originalText.Length > 2 && originalText[1] == '@');
+            var isVerbatim = (originalText[0] == '$' && originalText[1] == '@') ||
+                             (originalText[0] == '@' && originalText[1] == '$');
 
-            Debug.Assert(originalToken.Kind == SyntaxKind.InterpolatedStringToken);
+            // compute the positions of the interpolations in the original string literal, if there was an error or not,
+            // and where the close quote can be found.
             var interpolations = ArrayBuilder<Lexer.Interpolation>.GetInstance();
-            SyntaxDiagnosticInfo error;
-            bool closeQuoteMissing;
-            using (var tempLexer = new Lexer(Text.SourceText.From(originalText), this.Options, allowPreprocessorDirectives: false))
+
+            rescanInterpolation(out var openQuoteRange, out var error, out var closeQuoteRange);
+
+            var result = SyntaxFactory.InterpolatedStringExpression(
+                getOpenQuote(openQuoteRange), getContent(interpolations), getCloseQuote(closeQuoteRange));
+
+            interpolations.Free();
+            if (error != null)
+            {
+                result = result.WithDiagnosticsGreen(new[] { error });
+            }
+
+            Debug.Assert(originalToken.ToFullString() == result.ToFullString()); // yield from text equals yield from node
+            return CheckFeatureAvailability(result, MessageID.IDS_FeatureInterpolatedStrings);
+
+            void rescanInterpolation(out Range openQuoteRange, out SyntaxDiagnosticInfo error, out Range closeQuoteRange)
             {
-                // compute the positions of the interpolations in the original string literal, and also compute/preserve
-                // lexical errors
+                using var tempLexer = new Lexer(SourceText.From(originalText), this.Options, allowPreprocessorDirectives: false);
                 var info = default(Lexer.TokenInfo);
-                tempLexer.ScanInterpolatedStringLiteralTop(interpolations, isVerbatim, ref info, out error, out closeQuoteMissing);
+                tempLexer.ScanInterpolatedStringLiteralTop(ref info, out error, out openQuoteRange, interpolations, out closeQuoteRange);
             }
 
-            // Make a token for the open quote $" or $@" or @$"
-            var openQuoteIndex = isVerbatim ? 2 : 1;
-            Debug.Assert(originalText[openQuoteIndex] == '"');
-
-            var openQuoteKind = isVerbatim
-                ? SyntaxKind.InterpolatedVerbatimStringStartToken // $@ or @$
-                : SyntaxKind.InterpolatedStringStartToken; // $
-
-            var openQuoteText = isAltInterpolatedVerbatim
-                ? "@$\""
-                : isVerbatim
-                    ? "$@\""
-                    : "$\"";
```

Comment:
this part, in particular is just wonky.  We shoudl never be regenerating text.  We have the actual original source text, and can know the spans that things were at.  So we should always pull from that to ensure all data is consistent.